File: web/src/lib/google-maps.ts
const GEOCODING_URL = "https://maps.googleapis.com/maps/api/geocode/json";
const DISTANCE_MATRIX_URL = "https://maps.googleapis.com/maps/api/distancematrix/json";

function getApiKey(): string {
  const key = process.env.GOOGLE_MAPS_API_KEY;
  if (!key) throw new Error("GOOGLE_MAPS_API_KEY is not set");
  return key;
}

export interface GeocodeResult {
  lat: number;
  lng: number;
  formatted_address?: string;
}

/** Resolve a location string (e.g. "Sellery Hall", "Madison WI") to coordinates using Google Geocoding API. */
export async function resolveLocation(locationString: string): Promise<GeocodeResult | null> {
  const key = getApiKey();
  const res = await fetch(
    `${GEOCODING_URL}?address=${encodeURIComponent(locationString)}&key=${key}`
  );
  const data = await res.json();
  if (data.status !== "OK" || !data.results?.[0]) return null;
  const loc = data.results[0].geometry?.location;
  const formatted = data.results[0].formatted_address;
  if (!loc) return null;
  return {
    lat: loc.lat,
    lng: loc.lng,
    formatted_address: formatted,
  };
}

/**
 * Get distance in km between origin and destination using Google Distance Matrix API.
 * Returns the first element's distance value in km.
 */
export async function calculateDistanceKm(
  origin: { lat: number; lng: number },
  destination: { lat: number; lng: number }
): Promise<number | null> {
  const key = getApiKey();
  const orig = `${origin.lat},${origin.lng}`;
  const dest = `${destination.lat},${destination.lng}`;
  const res = await fetch(
    `${DISTANCE_MATRIX_URL}?origins=${encodeURIComponent(orig)}&destinations=${encodeURIComponent(dest)}&key=${key}`
  );
  const data = await res.json();
  if (data.status !== "OK" || !data.rows?.[0]?.elements?.[0]) return null;
  const el = data.rows[0].elements[0];
  if (el.status !== "OK" || !el.distance) return null;
  return el.distance.value / 1000; // meters -> km
}

/**
 * Get distances from one origin to many destinations in one call (more efficient).
 * Returns array of distances in km in same order as destinations.
 */
export async function calculateDistancesKm(
  origin: { lat: number; lng: number },
  destinations: Array<{ lat: number; lng: number }>
): Promise<(number | null)[]> {
  if (destinations.length === 0) return [];
  const key = getApiKey();
  const orig = `${origin.lat},${origin.lng}`;
  const destStr = destinations.map((d) => `${d.lat},${d.lng}`).join("|");
  const res = await fetch(
    `${DISTANCE_MATRIX_URL}?origins=${encodeURIComponent(orig)}&destinations=${encodeURIComponent(destStr)}&key=${key}`
  );
  const data = await res.json();
  if (data.status !== "OK" || !data.rows?.[0]?.elements) return destinations.map(() => null);
  return data.rows[0].elements.map((el: { status: string; distance?: { value: number } }) =>
    el.status === "OK" && el.distance ? el.distance.value / 1000 : null
  );
}

File: backend/app/google_maps.py
from typing import List, Optional
import httpx
from app.config import settings

GEOCODING_URL = "https://maps.googleapis.com/maps/api/geocode/json"
DISTANCE_MATRIX_URL = "https://maps.googleapis.com/maps/api/distancematrix/json"


def resolve_location(location_string: str) -> Optional[dict]:
    key = settings.google_maps_api_key
    if not key:
        return None
    with httpx.Client() as client:
        r = client.get(
            GEOCODING_URL,
            params={"address": location_string, "key": key},
        )
    if r.status_code != 200:
        return None
    data = r.json()
    if data.get("status") != "OK" or not data.get("results"):
        return None
    loc = data["results"][0].get("geometry", {}).get("location")
    if not loc:
        return None
    return {
        "lat": loc["lat"],
        "lng": loc["lng"],
        "formatted_address": data["results"][0].get("formatted_address"),
    }


def calculate_distances_km(
    origin: dict,
    destinations: List[dict],
) -> List[Optional[float]]:
    if not destinations:
        return []
    key = settings.google_maps_api_key
    if not key:
        return [None] * len(destinations)
    orig = f"{origin['lat']},{origin['lng']}"
    dest_str = "|".join(f"{d['lat']},{d['lng']}" for d in destinations)
    with httpx.Client() as client:
        r = client.get(
            DISTANCE_MATRIX_URL,
            params={"origins": orig, "destinations": dest_str, "key": key},
        )
    if r.status_code != 200:
        return [None] * len(destinations)
    data = r.json()
    if data.get("status") != "OK" or not data.get("rows"):
        return [None] * len(destinations)
    elements = data["rows"][0].get("elements", [])
    result = []
    for el in elements:
        if el.get("status") == "OK" and el.get("distance"):
            result.append(el["distance"]["value"] / 1000)
        else:
            result.append(None)
    return result
